File: sdks/kotlin/src/main/kotlin/dev/orion/sdk/runtime/RunEvent.kt
package dev.orion.sdk.runtime

import dev.orion.sdk.AgentEventKind

/** Ordered immutable observation emitted during a run. */
internal data class RunEvent(
    /** Identifier of the observed run. */
    public val runId: String,
    /** Zero-based monotonic run-local sequence number. */
    public val sequence: Long,
    /** Strongly typed event payload. */
    public val kind: AgentEventKind,
) : RunItem {

    init {
        require(runId.isNotBlank()) { "run identifier must not be blank" }
        require(sequence >= 0) { "run event sequence must not be negative" }
    }

}

File: sdks/javascript/src/runtime/run-event.ts
import type {ProtocolError} from "./protocol-error.js";

interface EventEnvelope<Type extends string, Data> {

  /** Run identifier shared by every event in one trace. */
  readonly runId: string;

  /** Zero-based monotonic event sequence. */
  readonly sequence: number;

  /** Stable lifecycle discriminator. */
  readonly type: Type;

  /** Event-specific immutable payload. */
  readonly data: Data;

}

/** Ordered lifecycle event emitted while an Orion run advances. */
export type RunEvent =
  | EventEnvelope<"run_started", {readonly agentId: string}>
  | EventEnvelope<"model_requested", {readonly turn: number; readonly provider: string; readonly model: string}>
  | EventEnvelope<"model_completed", {readonly turn: number; readonly output: string; readonly toolCallCount: number}>
  | EventEnvelope<"tool_requested" | "tool_completed", {readonly actionId: string; readonly callId: string; readonly name: string}>
  | EventEnvelope<"run_completed", {readonly output: string}>
  | EventEnvelope<"run_failed", {readonly error: ProtocolError}>
  | EventEnvelope<"run_cancelled", null>;
